src/main/scala/org/apache/spark/sql/kinesis/KinesisPosition.scala

```diff
@@ -53,3 +59,72 @@ class ShardEnd() extends KinesisPosition {
   override val iteratorType = "SHARD_END"
   override val iteratorPosition = ""
 }
+
+private[kinesis] object KinesisPosition {
+  def make(iteratorType: String, iteratorPosition: String): KinesisPosition = iteratorType match {
+    case iterType if "TRIM_HORIZON".equalsIgnoreCase(iterType) => new TrimHorizon()
+    case iterType if "LATEST".equalsIgnoreCase(iterType) => new Latest()
+    case iterType if "AT_TIMESTAMP".equalsIgnoreCase(iterType) => new AtTimeStamp(iteratorPosition)
+    case iterType if "AT_SEQUENCE_NUMBER".equalsIgnoreCase(iterType) =>
+      new AtSequenceNumber(iteratorPosition)
+    case iterType if "AFTER_SEQUENCE_NUMBER".equalsIgnoreCase(iterType) =>
+      new AfterSequenceNumber(iteratorPosition)
+    case iterType if "SHARD_END".equalsIgnoreCase(iterType) => new ShardEnd()
+  }
+}
+
+/**
+ * Specifies initial position in Kenesis to start read from on the application startup.
+ * @param shardPositions map of shardId->KinesisPosition
+ * @param defaultPosition position that is used for shard that is requested but not present in map
+ */
+private[kinesis] class InitialKinesisPosition(shardPositions: Map[String, KinesisPosition],
+                                              defaultPosition: KinesisPosition = new Latest())
```

Comment:
Using "Latest" as a default position might lead to data loss. 

Say we started fromCheckpointJson at time = 0 and Let's say resharding interval is 1 hr. Now if we split an existing shard. We will know about the next shard only at t = 1 hr. Since we would use "latest" as the initial position for the new shard. We would not process any data in the shard till time = 1 hr

Reply:
Oh, I see. Thank you for the explanation.
Removed a default value for default position, and altered KinesisSourceProvider as you've suggested.